File: Cora/Objective-C/Controllers/SettingsVC.m
#import "SettingsVC.h"


@implementation SettingsVC {

	UILabel *darwinLabel;
	UIButton *accentButton;
	UIView *indicatorView;
	CAShapeLayer *indicatorShapeLayer;
	UIStackView *stackView;
	UIButton *sourceCodeButton;
	UILabel *copyrightLabel;
	UILayoutGuide *containerGuide;
	UILayoutGuide *buttonContainerGuide;
	UIColorPickerViewController *pickerController;

}


- (id)initWithNibName:(NSString *)aNib bundle:(NSBundle *)aBundle {

	self = [super initWithNibName:aNib bundle:aBundle];

	if(self) {

		[[UserDefaultsManager sharedInstance] loadAccentColor];

		[self setupUI];
		[self setupPicker];

	}

	return self;

}


- (void)viewDidLayoutSubviews {

	[super viewDidLayoutSubviews];

	[self layoutUI];

}


- (void)viewWillAppear:(BOOL)animated {

	[super viewWillAppear:animated];

	[[UserDefaultsManager sharedInstance] setSwitchState];

}


- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {

	[super traitCollectionDidChange:previousTraitCollection];

	/*--- for some reason systemBackgroundColor is gray when presenting
	the view controller, and I want black so I do it here manually with
	both dark and light mode support :skull: ---*/

	self.view.backgroundColor = kUserInterfaceStyle ? UIColor.blackColor : UIColor.whiteColor;

}


- (void)setupUI {

	/*--- a table view clearly makes more sense here, but the switch state
		wouldn't be saved when the app is killed and re-opened. I figured it out
		eventually, but that was after I made all of this :skull:
		also I was picky and wanted to use layout guides. In the Swift version
		I used a table view ---*/

	containerGuide = [UILayoutGuide new];
	[self.view addLayoutGuide:containerGuide];

	buttonContainerGuide = [UILayoutGuide new];
	[self.view addLayoutGuide:buttonContainerGuide];

	darwinLabel = [UILabel new];
	darwinLabel.font = [UIFont fontWithName:@"Courier" size:15.5];
	darwinLabel.text = @"Print Darwin Information";
	darwinLabel.textColor = [ColorManager sharedInstance].accentColor;
	darwinLabel.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview:darwinLabel];

	commandSwitch = [UISwitch new];
	commandSwitch.onTintColor = [ColorManager sharedInstance].accentColor;
	commandSwitch.translatesAutoresizingMaskIntoConstraints = NO;
	[commandSwitch addTarget : self action:@selector(switchChanged:) forControlEvents:UIControlEventValueChanged];
	[self.view addSubview:commandSwitch];

	accentButton = [UIButton new];
	accentButton.titleLabel.font = [UIFont fontWithName:@"Courier" size:15.5];
	accentButton.translatesAutoresizingMaskIntoConstraints = NO;
	[accentButton setTitle : @"Change Accent Color" forState:UIControlStateNormal];
	[accentButton setTitleColor : [ColorManager sharedInstance].accentColor forState:UIControlStateNormal];
	[accentButton addTarget : self action:@selector(presentColorPickerController) forControlEvents:UIControlEventTouchUpInside];		
	[self.view addSubview:accentButton];

	indicatorView = [UIView new];
	indicatorView.clipsToBounds = YES;
	indicatorView.layer.borderColor = UIColor.whiteColor.CGColor;
	indicatorView.layer.borderWidth = 2;
	indicatorView.layer.cornerRadius = 15;
	indicatorView.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview:indicatorView];

	indicatorShapeLayer = [CAShapeLayer new];
	indicatorShapeLayer.path = [UIBezierPath bezierPathWithOvalInRect:CGRectMake(0,0,30,30)].CGPath;
	indicatorShapeLayer.fillColor = [ColorManager sharedInstance].accentColor.CGColor;
	[indicatorView.layer addSublayer:indicatorShapeLayer];

	stackView = [UIStackView new];
	stackView.axis = UILayoutConstraintAxisVertical;
	stackView.spacing = 10;
	stackView.alignment = UIStackViewAlignmentCenter;
	stackView.distribution = UIStackViewDistributionFill;
	stackView.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview:stackView];

	sourceCodeButton = [UIButton new];
	sourceCodeButton.alpha = 0.5;
	sourceCodeButton.titleLabel.font = [UIFont fontWithName:@"Courier" size:15.5];
	[sourceCodeButton setTitle : @"Source Code" forState:UIControlStateNormal];
	[sourceCodeButton setTitleColor : [ColorManager sharedInstance].accentColor forState:UIControlStateNormal];
	[sourceCodeButton addTarget : self action:@selector(didTapSourceCodeButton) forControlEvents:UIControlEventTouchUpInside];		
	[stackView addArrangedSubview:sourceCodeButton];

	copyrightLabel = [UILabel new];
	copyrightLabel.font = [UIFont fontWithName:@"Courier" size:10];
	copyrightLabel.text = @"2021 © Luki120";
	copyrightLabel.alpha = 0.5;
	copyrightLabel.textColor = [ColorManager sharedInstance].accentColor;
	copyrightLabel.textAlignment = NSTextAlignmentCenter;
	[stackView addArrangedSubview:copyrightLabel];

}


- (void)setupPicker {

	pickerController = [UIColorPickerViewController new];
	pickerController.delegate = self;
	pickerController.selectedColor = [ColorManager sharedInstance].accentColor;
	pickerController.modalPresentationStyle = UIModalPresentationPopover;

}


- (void)layoutUI {

	UILayoutGuide *guide = self.view.safeAreaLayoutGuide;

	[containerGuide.topAnchor constraintEqualToAnchor : self.view.topAnchor].active = YES;
	[containerGuide.leadingAnchor constraintEqualToAnchor : guide.leadingAnchor].active = YES;
	[containerGuide.trailingAnchor constraintEqualToAnchor : guide.trailingAnchor].active = YES;
	[containerGuide.heightAnchor constraintEqualToConstant : 44].active = YES;

	[buttonContainerGuide.topAnchor constraintEqualToAnchor : containerGuide.bottomAnchor constant : 30].active = YES;
	[buttonContainerGuide.leadingAnchor constraintEqualToAnchor : guide.leadingAnchor].active = YES;
	[buttonContainerGuide.trailingAnchor constraintEqualToAnchor : guide.trailingAnchor].active = YES;
	[buttonContainerGuide.heightAnchor constraintEqualToConstant : 44].active = YES;

	[darwinLabel.centerYAnchor constraintEqualToAnchor : commandSwitch.centerYAnchor].active = YES;
	[darwinLabel.leadingAnchor constraintEqualToAnchor : containerGuide.leadingAnchor constant : 20].active = YES;

	[commandSwitch.topAnchor constraintEqualToAnchor : containerGuide.topAnchor constant : 30].active = YES;
	[commandSwitch.trailingAnchor constraintEqualToAnchor : containerGuide.trailingAnchor constant : -20].active = YES;

	[accentButton.centerYAnchor constraintEqualToAnchor : indicatorView.centerYAnchor].active = YES;
	[accentButton.leadingAnchor constraintEqualToAnchor : buttonContainerGuide.leadingAnchor constant : 20].active = YES;

	[indicatorView.centerYAnchor constraintEqualToAnchor : buttonContainerGuide.centerYAnchor].active = YES;
	[indicatorView.trailingAnchor constraintEqualToAnchor : buttonContainerGuide.trailingAnchor constant: -28].active = YES;
	[indicatorView.widthAnchor constraintEqualToConstant : 30].active = YES;
	[indicatorView.heightAnchor constraintEqualToConstant : 30].active = YES;

	[stackView.centerXAnchor constraintEqualToAnchor : self.view.centerXAnchor].active = YES;
	[stackView.topAnchor constraintEqualToAnchor : buttonContainerGuide.bottomAnchor constant : 50].active = YES;

}


- (void)switchChanged:(UISwitch *)sender {

	[[UserDefaultsManager sharedInstance] saveSwitchState];

	[NSNotificationCenter.defaultCenter postNotificationName:@"launchChosenTask" object:nil];

}


- (void)presentColorPickerController {

	[self presentViewController:pickerController animated:YES completion:nil];

}


- (void)colorPickerViewControllerDidSelectColor:(UIColorPickerViewController *)colorPicker {

	[ColorManager sharedInstance].accentColor = colorPicker.selectedColor;

	darwinLabel.textColor = [ColorManager sharedInstance].accentColor;
	commandSwitch.onTintColor = [ColorManager sharedInstance].accentColor;

	[accentButton setTitleColor : [ColorManager sharedInstance].accentColor forState:UIControlStateNormal];
	indicatorShapeLayer.fillColor = [ColorManager sharedInstance].accentColor.CGColor;

	[sourceCodeButton setTitleColor : [ColorManager sharedInstance].accentColor forState:UIControlStateNormal];
	copyrightLabel.textColor = [ColorManager sharedInstance].accentColor;

	self.rootVC.darwinLabel.textColor = [ColorManager sharedInstance].accentColor;
	self.rootVC.uptimeLabel.textColor = [ColorManager sharedInstance].accentColor;
	self.rootVC.settingsButton.tintColor = [ColorManager sharedInstance].accentColor;

	[[UserDefaultsManager sharedInstance] saveAccentColor];

}


- (void)didTapSourceCodeButton {

	NSURL *url = [NSURL URLWithString:@"https://github.com/Luki120/iOS-Apps/tree/main/Cora"];

	SFSafariViewController *safariViewController = [[SFSafariViewController alloc] initWithURL:url];
	safariViewController.delegate = self;
	[self presentViewController:safariViewController animated:YES completion:nil];

}


@end

File: Released/Cora/Swift/UIKit/Controllers/SettingsVC.swift
import UIKit
import SafariServices


final class SettingsVC: UITableViewController {

	private let accentLabel: UILabel = {
		let label = UILabel()
		label.font = UIFont(name: "Courier", size: 15)
		label.text = "Change Accent Color"
		label.textAlignment = .center
		label.numberOfLines = 0
		label.translatesAutoresizingMaskIntoConstraints	= false
		return label
	}()

	private let darwinLabel: UILabel = {
		let label = UILabel()
		label.font = UIFont(name: "Courier", size: 15)
		label.text = "Print Darwin Information" 
		label.textAlignment = .center
		label.numberOfLines = 0
		label.translatesAutoresizingMaskIntoConstraints = false
		return label
	}()

	private let indicatorView: UIView = {
		let view = UIView()
		view.clipsToBounds = true
		view.layer.borderColor = UIColor.white.cgColor
		view.layer.borderWidth = 2
		view.layer.cornerRadius = 15
		view.translatesAutoresizingMaskIntoConstraints = false
		return view
	}()

	private let indicatorShapeLayer: CAShapeLayer = {
		let shapeLayer = CAShapeLayer()
		shapeLayer.path = UIBezierPath(ovalIn: CGRect(x: 0, y: 0, width: 30, height: 30)).cgPath
		return shapeLayer
	}()

	private let stackView: UIStackView = {
		let stackView = UIStackView()
		stackView.axis = .vertical
		stackView.spacing = 10
		stackView.alignment = .center
		stackView.distribution = .fill
		stackView.translatesAutoresizingMaskIntoConstraints = false
		return stackView
	}()

	private let sourceCodeButton: UIButton = {
		let button = UIButton()
		button.alpha = 0.5
		button.titleLabel?.font = UIFont(name: "Courier", size: 15.5)
		button.setTitle("Source Code", for: .normal)
		button.addTarget(self, action: #selector(didTapSourceCodeButton), for: .touchUpInside)
		return button
	}()

	private let copyrightLabel: UILabel = {
		let label = UILabel()
		label.alpha = 0.5
		label.font = UIFont(name: "Courier", size: 10)
		label.text = "2022 © Luki120"
		label.textAlignment = .center
		return label
	}()

	private let uptRootVC = UPTRootVC()
	private let colorPicker = UIColorPickerViewController()

	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}

	init() {

		super.init(nibName: nil, bundle: nil)

		setupUI()

		tableView.register(UITableViewCell.self, forCellReuseIdentifier: "VanillaCell")

		GlobalManager.sharedInstance.commandSwitch.addTarget(self, action: #selector(switchStateChanged), for: .valueChanged)

	}

	override func viewWillAppear(_ animated: Bool) {

		super.viewWillAppear(animated)
		UserDefaultsManager.sharedInstance.setSwitchState()

	}

	override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {

		super.traitCollectionDidChange(previousTraitCollection)
		tableView.backgroundColor = traitCollection.userInterfaceStyle == .dark ? .black : .white

	}

	private func setupUI() {

		UserDefaultsManager.sharedInstance.loadAccentColor()

		darwinLabel.textColor = ColorManager.sharedInstance.accentColor
		GlobalManager.sharedInstance.commandSwitch.onTintColor = ColorManager.sharedInstance.accentColor

		accentLabel.textColor = ColorManager.sharedInstance.accentColor
		indicatorShapeLayer.fillColor = ColorManager.sharedInstance.accentColor.cgColor

		sourceCodeButton.setTitleColor(ColorManager.sharedInstance.accentColor, for: .normal)
		copyrightLabel.textColor = ColorManager.sharedInstance.accentColor

		tableView.separatorStyle = .none

		setupPicker()
		setupStackView()

	}

	private func setupPicker() {

		colorPicker.delegate = self
		colorPicker.selectedColor = ColorManager.sharedInstance.accentColor
		colorPicker.modalPresentationStyle = .popover

	}

	private func setupStackView() {

		view.addSubview(stackView)
		stackView.addArrangedSubview(sourceCodeButton)
		stackView.addArrangedSubview(copyrightLabel)

		let guide = view.safeAreaLayoutGuide

		stackView.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
		stackView.bottomAnchor.constraint(equalTo: guide.bottomAnchor, constant: -20).isActive = true

	}

	// MARK: - Actions

	@objc private func switchStateChanged() {

		UserDefaultsManager.sharedInstance.saveSwitchState()

		NotificationCenter.default.post(name: Notification.Name("launchChosenTask"), object: nil)

	}

	@objc private func didTapSourceCodeButton() {

		let url = URL(string: "https://github.com/Luki120/iOS-Apps/tree/main/Cora")

		guard let theURL = url else {
			return
		}

		let safariVC = SFSafariViewController(url: theURL)
		safariVC.delegate = self
		safariVC.modalPresentationStyle = .pageSheet
		present(safariVC, animated: true, completion: nil)

	}

}


extension SettingsVC {

	// MARK: - Table View Data Source

	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {

		return 2

	}

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

		let cell = tableView.dequeueReusableCell(withIdentifier: "VanillaCell", for: indexPath)

		cell.selectionStyle = .none
		cell.backgroundColor = .clear

		switch indexPath.row {

			case 0:

				cell.contentView.addSubview(darwinLabel)

				cell.accessoryView = GlobalManager.sharedInstance.commandSwitch

				darwinLabel.centerYAnchor.constraint(equalTo: cell.contentView.centerYAnchor).isActive = true
				darwinLabel.leadingAnchor.constraint(equalTo: cell.contentView.leadingAnchor, constant: 20).isActive = true

			case 1:

				cell.contentView.addSubview(accentLabel)

				accentLabel.centerYAnchor.constraint(equalTo: cell.contentView.centerYAnchor).isActive = true
				accentLabel.leadingAnchor.constraint(equalTo: cell.contentView.leadingAnchor, constant: 20).isActive = true

				cell.contentView.addSubview(indicatorView)
				indicatorView.layer.addSublayer(indicatorShapeLayer)

				indicatorView.centerYAnchor.constraint(equalTo: cell.contentView.centerYAnchor).isActive = true
				indicatorView.trailingAnchor.constraint(equalTo: cell.contentView.trailingAnchor, constant: -28).isActive = true
				indicatorView.widthAnchor.constraint(equalToConstant: 30).isActive = true
				indicatorView.heightAnchor.constraint(equalToConstant: 30).isActive = true

			default: break

		}

		return cell

	}

}


extension SettingsVC {

	// MARK: - Table View Delegate

	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {

		tableView.deselectRow(at: indexPath, animated: true)

		if indexPath.row == 1 { present(colorPicker, animated: true, completion: nil) }

	}

	override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {

		let headerView = UIView()
		return headerView

	}

}


extension SettingsVC: SFSafariViewControllerDelegate, UIColorPickerViewControllerDelegate {

	// MARK: - UIColorPickerViewControllerDelegate

	func colorPickerViewControllerDidSelectColor(_ viewController: UIColorPickerViewController) {

		ColorManager.sharedInstance.accentColor = viewController.selectedColor

		darwinLabel.textColor = ColorManager.sharedInstance.accentColor
		GlobalManager.sharedInstance.commandSwitch.onTintColor = ColorManager.sharedInstance.accentColor
		accentLabel.textColor = ColorManager.sharedInstance.accentColor

		indicatorShapeLayer.fillColor = ColorManager.sharedInstance.accentColor.cgColor

		sourceCodeButton.setTitleColor(ColorManager.sharedInstance.accentColor, for: .normal)
		copyrightLabel.textColor = ColorManager.sharedInstance.accentColor

		uptRootVC.darwinLabel.textColor = ColorManager.sharedInstance.accentColor
		uptRootVC.uptimeLabel.textColor = ColorManager.sharedInstance.accentColor
		uptRootVC.settingsButton.tintColor = ColorManager.sharedInstance.accentColor	

		NotificationCenter.default.post(name: Notification.Name("updateAccentColors"), object: nil)

		UserDefaultsManager.sharedInstance.saveAccentColor()

	}

}
